In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import os

In [27]:
def print_reviews(df, n, seed=42):
    np.random.seed(seed)

    print("Bad reviews")
    for i in range(0, n):
        # Randomly sample from bottom 5% to bottom 15%
        id = np.random.randint(int(len(df)*0.05), int(len(df)*0.15))
        print(f"{id}\t{df.text.iloc[id]}")

    print("\nGood reviews")
    for i in range(1, n + 1):
        print(f"{-i}\t{df.text.iloc[-i]}")

    print("\nRandom reviews")
    for _ in range(0, n):
        id = np.random.randint(int(len(df)*0.15), int(len(df)*0.75))
        print(f"{id}\t{df.text.iloc[id]}")

In [3]:
path_to_data = './../../../data'

## RateBeer


In [4]:
path = os.path.join(path_to_data, 'RateBeer')

df_scores = pd.read_pickle(os.path.join(path, 'rev_w_scores.pkl'))
df_reviews = pd.read_csv(os.path.join(path, 'ratings_RB_clean.csv'))

In [5]:
df_reviews = df_reviews[df_reviews['lang_tag'] == 'en']

In [25]:
top3_beer_id = df_scores.beer_id.value_counts().head(3).index.to_list()

guinness = df_scores[df_scores['beer_id'] == top3_beer_id[0]]
rochefort = df_scores[df_scores['beer_id'] == top3_beer_id[1]]
chimay = df_scores[df_scores['beer_id'] == top3_beer_id[2]]

n_rev = 5

In [12]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[0]], guinness, how='outer', on='user_id')

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Guinness Draught
Bad reviews
343	Very few things in life are as consistently good as a Guiness.  Until the day when I can get to Ireland, I’l drink it here in the states every time I do a part in a Shakespeare festival.
676	This beer is completely overrated.  First off, it is only drinkable by tap, the can and bottle versions aren't that great.  It's not that bad in the beginning because the flavor is decent, but the palate is too heavy, and by the end of the pint, the flavor is tiring.
589	One of the few beers that lives up to the hype around it. Good color. Nice head. Overall a solid buy.
511	This beer, like any beer is much better on tap than out of the bottle or can. It fills you up. It’s like a meal in a glass.

Good reviews
-1	14.9 ounce can into imperial pint glass, canned on 5/26/2016. Pours crystal clear very dark brown color with a 1 finger dense and creamy light tan head that lasts the whole glass. Thick creamy lacing clings down the glass. Fantastic appearance. Aromas o

In [15]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[1]], rochefort, how='outer', on='user_id')

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Rochefort Trappistes 10
Bad reviews
340	Very strong alcohol taste. I prefer the 8 or 6 over the 10. Compared to other Trappist beers, this one ranks very high.
673	This was a very nice belgian strong ale.  A very deep and dark color.  Lots of little flavors seep through the thickness of this brew.  Nicely done.
586	Unique style of Trappists, caramel and prunes, only sad point: watch out for the huge bottle deposit when served out of the house..
508	To my fellow Vets on this Veterans Day 2011 - SALUTE!! My gosh, brews like this have me wishing I was born in Belgium! I do believe I have found the sixth beer to go into my top-5 favorites list!!! This is definitely part of my winter holiday food and drink repertoire! Some hearty smoked cheese and smoked wild game (venison, wild boar) are definitely going to be paired up with this hearty brew. I could see this being my go-to brew on those cold wintry night by the fireside and a hearty cigar. The alcohol content (11.3% ABV) is absolutely war

In [16]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[2]], chimay, how='outer', on='user_id')

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Chimay Bleue (Blue) / Grande Réserve
Bad reviews
321	Tasted a bottle of 1999. First taste reminds more of old books and dust but in the end a subtle sweet taste takes the upperhand. Very nice carbonisation for this age still.
654	A trappist beer. Dark nice taste of dark choclate.It dossen’t fell like it’s 9%, one hellluva nice beer from Belgium.It’s easy to get but its nice
567	3 liter bottleThis is a yeast driven malty beer. It went very well with our Thanksgiving dinner. A solid beer, worth revisiting from time to time.
489	Great beer with a small touch of alcohol in the aroma.  Just enough alcohol to warm the senses and a wonderful taste to match.

Good reviews
-1	(Re-rated on 3/8/2016):Rationale: I was new to the style back then. With experience I want to revisit one of my favorite, and see if I like it as much as I was then. So here goes!330 ml bottle bought at CitySuper in Harbour City, Tsim Sha Tsui.Looks: Cloudy hazy amber brown colour with a layer of yeast sediments; One

### Rochefort Trappistes 10 

Last, but certainly not least, is their Belgian Quad, the Trappistes Rochefort 10. It had a dark ruby color with a lasting tan head. The aroma with very rich with a strong maltiness and sweet molasses coming through. The flavor followed the aroma with a strong and rich malty body with notes of dark fruit. There is a touch of sweetness but it manages to stay on the dry side. The 11.3% is masked well, but it is still warming. The body is smooth and silky, and it has a dry, malty finish. This is one of the highest-ranked Belgian Quads for a reason.

[link](https://thebeerjunkies.com/trappistes-rochefort/)

### Guinness Draught

"Swirling clouds tumble as the storm begins to calm, settle, breathe in the moment, then break through the smooth, light head to the bittersweet reward. Unmistakably Guinness from the first velvet sip to the last, lingering drop and every satisfying mouthful in between. Pure beauty. Pure Guinness." - Guinness

Aroma: “Sweet dark fruit character. Cherry, raisin, and roasty malt character with slight cocoa, chocolate, and coffee flavor. A little sourness and subtle vanilla bean notes.”

Flavor: “Light bodied with some slight creaminess and sweetness giving way to fairly aggressive bittering hops. Roasty, sharp astringent character. Some estery fruit characteristics. Creaminess carries through in the finish. Finishes dry with slightly oxidized notes.”

Overall: “The nose promised more than the beer provided. Somewhat light bodied with a roast character that dominates and a bit of creaminess to balance. Tannic harshness is too high—could use a bit less bittering or a touch more body to mellow it out. But the more you drink it, the more you enjoy it.”

[link](https://beerandbrewing.com/review/guinness-draught)

### Chimay Bleue (Blue) - Grande Réserve 

(Deepl translated) [link](https://www.vandb.fr/biere/chimay-bleue-grande-reserve-75cl.html)

Deep brown colour, creamy beige mousse

Aromas of candied fruit, plums, caramel, spices

Powerful, rich and creamy texture, notes of caramel, brown sugar, toasted malt, candied fruit, light hints of alcohol

On tasting, Chimay Bleue Grande Réserve has a deep brown colour covered with a creamy beige head. Its symphony of aromas tickles the senses, revealing notes of candied fruit, plums, caramel and spices. A beautiful Dark Ale that's well worth the diversions!



### Black and Tan

Particularly "nice" beer, used in other examples

In [29]:
# black and tan with most reviews
black_and_tan_id = 837

black_and_tan = df_scores[df_scores['beer_id'] == black_and_tan_id]

df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == black_and_tan_id], black_and_tan, how='outer', on='user_id')

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev, 11)

Mississippi Mud Black & Tan
Bad reviews
64	This stuff is great! It’s one those beers that tastes good warm.
102	Mild malty aroma, from the smell i thought it was going to taste really good, pretty mild and boring though. It does have a neat looking bodel atleast.
94	had never ahd this and was in a crappy store in a great beer town (Denver) but this was the only single even interesting.  it’s blah; but better than the overall score.  it is what it says it is a porter watered down by a pils.  kinda of a weak brown ale but no issues.  at least average beer
52	The name is appropriate. Sort of a one time ’gag’ beer.
115	This was my first dark beer and I loved it.  That was a long time ago, so I will have to try it again.

Good reviews
-1	16 ounce can into pint glass, canned on 1/22/2015. Pours slightly hazy reddish brown color with a 2 finger dense and fluffy light khaki head with good retention, that reduces to a small cap that lingers. Light spotty soapy lacing clings down the gla

In [ ]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[0]], guinness, how='outer', on='user_id')

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

In [36]:
del df_scores, df_reviews, df_merged

## BeerAdvocate


In [ ]:
path = os.path.join(path_to_data, 'BeerAdvocate')

df_scores = pd.read_pickle(os.path.join(path, 'rev_w_scores.pkl'))
df_reviews = pd.read_csv(os.path.join(path, 'ratings_BA_clean.csv'))

df_reviews = df_reviews[df_reviews['lang_tag'] == 'en']

FileNotFoundError: [Errno 2] No such file or directory: './../../../data/.BeerAdvocate/rev_w_scores.pkl'

In [32]:
top3_beer_id = df_scores.beer_id.value_counts().head(3).index.to_list()

minuteIPA = df_scores[df_scores['beer_id'] == top3_beer_id[0]]
breakfast_stout = df_scores[df_scores['beer_id'] == top3_beer_id[1]]
pliny = df_scores[df_scores['beer_id'] == top3_beer_id[2]]

n_rev = 4

In [33]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[0]], minuteIPA, how='outer', on='user_id').dropna()

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Guinness Draught
Bad reviews
343	Very few things in life are as consistently good as a Guiness.  Until the day when I can get to Ireland, I’l drink it here in the states every time I do a part in a Shakespeare festival.
676	This beer is completely overrated.  First off, it is only drinkable by tap, the can and bottle versions aren't that great.  It's not that bad in the beginning because the flavor is decent, but the palate is too heavy, and by the end of the pint, the flavor is tiring.
589	One of the few beers that lives up to the hype around it. Good color. Nice head. Overall a solid buy.
511	This beer, like any beer is much better on tap than out of the bottle or can. It fills you up. It’s like a meal in a glass.

Good reviews
-1	14.9 ounce can into imperial pint glass, canned on 5/26/2016. Pours crystal clear very dark brown color with a 1 finger dense and creamy light tan head that lasts the whole glass. Thick creamy lacing clings down the glass. Fantastic appearance. Aromas o

In [34]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[1]], breakfast_stout, how='outer', on='user_id').dropna()

df_merged.sort_values(by=['expertness_score'], inplace=True)


print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Rochefort Trappistes 10
Bad reviews
340	Very strong alcohol taste. I prefer the 8 or 6 over the 10. Compared to other Trappist beers, this one ranks very high.
673	This was a very nice belgian strong ale.  A very deep and dark color.  Lots of little flavors seep through the thickness of this brew.  Nicely done.
586	Unique style of Trappists, caramel and prunes, only sad point: watch out for the huge bottle deposit when served out of the house..
508	To my fellow Vets on this Veterans Day 2011 - SALUTE!! My gosh, brews like this have me wishing I was born in Belgium! I do believe I have found the sixth beer to go into my top-5 favorites list!!! This is definitely part of my winter holiday food and drink repertoire! Some hearty smoked cheese and smoked wild game (venison, wild boar) are definitely going to be paired up with this hearty brew. I could see this being my go-to brew on those cold wintry night by the fireside and a hearty cigar. The alcohol content (11.3% ABV) is absolutely war

In [35]:
df_merged = pd.merge(df_reviews[df_reviews['beer_id'] == top3_beer_id[2]], pliny, how='outer', on='user_id').dropna()

df_merged.sort_values(by=['expertness_score'], inplace=True)

print(df_merged.beer_name_x[0])
print_reviews(df_merged, n_rev)

Chimay Bleue (Blue) / Grande Réserve
Bad reviews
321	Tasted a bottle of 1999. First taste reminds more of old books and dust but in the end a subtle sweet taste takes the upperhand. Very nice carbonisation for this age still.
654	A trappist beer. Dark nice taste of dark choclate.It dossen’t fell like it’s 9%, one hellluva nice beer from Belgium.It’s easy to get but its nice
567	3 liter bottleThis is a yeast driven malty beer. It went very well with our Thanksgiving dinner. A solid beer, worth revisiting from time to time.
489	Great beer with a small touch of alcohol in the aroma.  Just enough alcohol to warm the senses and a wonderful taste to match.

Good reviews
-1	(Re-rated on 3/8/2016):Rationale: I was new to the style back then. With experience I want to revisit one of my favorite, and see if I like it as much as I was then. So here goes!330 ml bottle bought at CitySuper in Harbour City, Tsim Sha Tsui.Looks: Cloudy hazy amber brown colour with a layer of yeast sediments; One

### 90 Minute IPA

[link](https://www.dogfish.com/brewery/beer/90-minute-ipa)<br>
With rich pine and fruity citrus hop aromas and a strong malt backbone, 90 Minute IPA created pungent, unapologetic flavor that led Esquire to call it “perhaps the best IPA in America.”<br>
Color: Golden amber <br>
Aroma: Heavy piney and citrus hop aromas <br>
Flavour: Brandied fruitcake, raisiney, citrusy <br>
Mouthfeel: Full and malty with slightly lingering bitterness

### Founders Breakfast Stout

[link](https://foundersbrewing.com/our-beer/breakfast-stout/)<br>
Coffee lovers: get your mugs ready because this is the perfect beer for you. Brewed with an abundance of flaked oats, specialty chocolates, and two types of coffee, this stout has an intense fresh-roasted java nose topped with a frothy, cinnamon-colored head that goes forever. With low bitterness and just the right amount of sweetness, you’ll be coming back for seconds.

Tasting notes: Breakfast-inspired flavors of roasted coffee, chocolate, and caramelized sugar, with low bitterness and just the right amount of sweetness.

### Pliny The Elder

[link](https://beerandbrewing.com/review/pliny-the-elder/)
Pliny the Elder is brewed with Amarillo, Centennial, CTZ, and Simcoe hops. It is well-balanced with malt, hops, and alcohol, slightly bitter with a fresh hop aroma of floral, citrus, and pine.

Aroma: “Nice bright tropical fruit with notes of mango, pineapple, grapefruit, and pine. Some noticeable light malt sweetness as it warms. Dank aroma.”

Flavor: “Nice malt sweetness that balances out the bitterness. Slight grassiness, notes of pineapple, pear, grapefruit, mango, and tangerine. Finish is decidedly bitter and lingering.”

Overall: “Good level of hop flavor and a delicately restrained bitterness to help showcase hop aroma and flavor.”

### Hopslam Ale  

[link](https://bellsbeer.com/beers/hopslam-ale/)<br>
Flavor Profile: Huge citrus, Floral & Fruit Hop Notes, Grapefruit, Stone Fruit, and Floral Aromas all from Hops
Beer Info: Selected specifically because of their aromatic qualities, these Pacific Northwest varieties contribute a pungent blend of grapefruit, stone fruit, and floral notes. A generous malt bill and a solid dollop of honey provide just enough body to keep the balance in check, resulting in a remarkably drinkable rendition of the Double IPA style.

In [ ]:
del df_scores, df_reviews, df_merged